daemon/logic/worklog.go

```diff
@@ -242,17 +270,257 @@ func (h *inviteApproveHandler) resolve(ctx context.Context, n *observer.Notifier
 	if err != nil {
 		return nil, err
 	}
+
+	return &apitypes.WorklogResult{
+		ID:      item.ID,
+		State:   apitypes.SuccessWorklogResult,
+		Message: "User invite approved and finalized.",
+	}, nil
+}
+
+type keyringMembersHandler struct {
+	engine *Engine
+}
+
+func (keyringMembersHandler) resolveErr() string {
+	return "Error adding user(s) to keyring"
+}
+
+func (h *keyringMembersHandler) list(ctx context.Context, org *envelope.Unsigned) ([]apitypes.WorklogItem, error) {
+	projects, err := h.engine.client.Projects.List(ctx, org.ID)
 	if err != nil {
-		return &apitypes.WorklogResult{
-			ID:      item.ID,
-			State:   apitypes.ErrorWorklogResult,
-			Message: "Error approving invite: " + err.Error(),
-		}, nil
+		return nil, err
+	}
+
+	// Find all of the credential graphs in this org.
+	// Search won't return all of the keyring memberships, so we have to
+	// use it just to find all of the distinct pathexps for keyrings.
+	//
+	// The paths map takes care of eliminating multiple versions of the keyring;
+	// the subsequent List call will return all versions.
+	cgs := newCredentialGraphSet()
+	orgName := org.Body.(*primitive.Org).Name
+	paths := make(map[string]*pathexp.PathExp)
+	for _, project := range projects {
+		projName := project.Body.(*primitive.Project).Name
+		graphs, err := h.engine.client.CredentialGraph.Search(ctx,
+			"/"+orgName+"/"+projName+"/*/*/*/*", h.engine.session.AuthID())
+		if err != nil {
+			return nil, err
+		}
+
+		for _, graph := range graphs {
+			var path *pathexp.PathExp
+			keyring := graph.GetKeyring()
+			switch keyring.Version {
+			case 1:
+				path = keyring.Body.(*primitive.KeyringV1).PathExp
+			case 2:
+				path = keyring.Body.(*primitive.Keyring).PathExp
+			}
+
+			paths[path.String()] = path
+		}
+	}
+
+	for _, pe := range paths {
+		graphs, err := h.engine.client.CredentialGraph.List(ctx, "", pe, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		err = cgs.Add(graphs...)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// To find out if any keyrings are missing members, we need walk through all
+	// active versions of each keyring, to see if anyone is missing.
+	// Inactive versions don't matter, as there is nothing there a user would
+	// want to access.
+	graphs, err := cgs.Active()
+	if err != nil {
+		return nil, err
+	}
+
+	// XXX: this logic will be much different when we selectively encode the
+	// members of a keyring based on ACLs.
+	members, err := getKeyringMembers(ctx, h.engine.client, org.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	missing := make(map[string]apitypes.WorklogItem)
+	for _, graph := range graphs {
+		for _, member := range members {
+			m, _, err := graph.FindMember(&member)
```

Comment:
(If you have a valid keypair but the currently encoded key is revoked, we should encode your new keys)